added: mic gain

msp1974 opened the pull request:
## In this PR

- Added microphone gain control feature
  - This provides a slider on the Hassmic app screen to set the gain from 1 to 11 (as per Spinal Tap!)
  - The gain is set to between 1 and approx 92x based on the formula `gain * (gain * 0.6)` to allow more sensitivity at lower values
  - The gain is set dynamically and hassmic does not need to be restarted for it to take effect

## Testing

It has been tested on a Lenovo SD10 set to 8 (gain of approx 50x)  and this delivers good pickup from upto 3 meters away.  Set to 11, you can whisper to it from about 1 meter away.

jeffc commented on app/app/wyoming.ts:
```diff
@@ -12,7 +17,7 @@ import {
 import {CheyenneSocket} from './cheyenne';
 import {DeviceEventEmitter} from 'react-native';
-import {UUIDManager} from './util';
+import AsyncStorage from '@react-native-async-storage/async-storage';
```
I think it makes more sense to make this a field in the settings proto rather than creating a new storage key directly.

If you haven't already (and I don't think I documented this anywhere yet, so that's my bad) - installing [pre-commit](https://pre-commit.com/) will automatically rebuild the generated proto files and do some other useful cleanup whenever you make a commit. You can also just run `./scripts/proto_regen.sh` after you change the `.proto` files to regenerate all of the client code.

msp1974 replied:
OK, let me look at that and move it into a proto setting and HA UI config.

msp1974 commented:
OK, forgive me if I have got this wrong.  Struggled to get my head around betterproto but think I did.  Let me know if I have done anything stupid!  Thx

jeffc commented on app/app/wyoming.ts:
```diff
@@ -1,7 +1,12 @@
+import {
+  APP_VERSION,
+  AUDIO_INFO,
+  STORAGE_MIC_GAIN,
```
not using this anymore; we can delete it

msp1974 replied:
removed

jeffc commented on app/app/constants.ts:
```diff
@@ -1,9 +1,10 @@
 export const HASSMIC_PORT = 11700;
 export const WYOMING_PORT = 10700;
 export const STORAGE_KEY_UUID = 'zeroconfuuid';
+export const STORAGE_MIC_GAIN = 'micgain';
```
this isn't used anymore; we can delete it

msp1974 replied:
removed

jeffc commented on app/app/wyoming.ts:
```diff
@@ -701,6 +705,9 @@ class WyomingServer_ {
         );
       }
```
The "settings changed" callback is meant to update the internal state of the wyoming server any time something in the settings blob is changed. we should probably add mic gain here too.

msp1974 replied:
removed these and used settings changed callback

jeffc commented on app/app/wyoming.ts:
```diff
@@ -712,13 +719,74 @@ class WyomingServer_ {
+  setMicGain = async (gain: number) => {
+    this.micGain = gain;
+    await Settings.setMicGain(gain);
+  };
```
Are these functions actually necessary?  If the mic gain gets loaded from settings in the constructor and updated in the settingschanged callback, we can just do Settings.setMicGain() from anywhere and it will stay up to date here.

msp1974 replied:
Removed and used settings changed callback

jeffc commented on app/app/main.tsx:
```diff
@@ -138,8 +140,18 @@ export default function Index() {
+  useEffect(() => {
+    WyomingServer.setMicGain(micGain)
+  }, [micGain]);
+
```
See my comment in wyoming.ts below, but I think we can get rid of the `WyomingServer.(set|load)MicGain()` in favor of using the `Settings` object directly and updating Wyoming in the callback. This will let any other consumers of the mic gain stay in sync

msp1974 replied:
Removed all this and made it a HASS UI controlled setting as I think that is more how you want this to work.

jeffc commented:
> OK, forgive me if I have got this wrong. Struggled to get my head around betterproto but think I did. Let me know if I have done anything stupid! Thx

It looks like you nailed it! If you haven't seen them before, [protocol buffers](https://protobuf.dev/) are pretty neat - you define the structure for some data, and the library takes care of generating the language-specific implementations of that datatype *and* the serialization that lets you pass instances of that structured data between languages. The idea is that I can define structured messages for storing data (like settings) or passing data (like info between the client and server) and not have to do any work to keep those definitions in sync between different parts of the code. Protocol buffers also provide some level of graceful backwards compatibility in case versions get out of sync.

I left some more comments that I hadn't noticed the first time (or that weren't relevant / are about new code). I hope my intentions are clear, but definitely feel free to ask questions or challenge my assertions/design decisions if you think there's a better way :)

Thanks again for all of your contributions!

msp1974 commented:
Yes that's all good.  I did do it this way at first but I couldn't get the view to refresh the gain value when it changed via the slider, so I did this sort of hacky thing!

Let me look again. My first attempt at RN, so happy with any pointers.

msp1974 commented:
Right.  I think I have this now as you intend it to work.  I have moved everything to use the settings changed callbacks (which works fine so not sure what I did wrong first time when I couldn't get them to work!).

I have also removed the on screen slider and made it a HASS UI setting - again, I think this is more how you want it to work.

Let me know....thx

jeffc commented on app/app/constants.ts:
```diff
@@ -3,7 +3,7 @@ export const WYOMING_PORT = 10700;
 export const STORAGE_KEY_RUN_BACKGROUND_TASK = 'bgtaskshouldrun';
 export const STORAGE_KEY_SAVED_SETTINGS_PROTO = 'savedsettingsproto';
-export const APP_VERSION = '0.9.4';
+export const APP_VERSION = '0.9.5';
```
Can you change the `VERSION` file and then run `scripts/update_vresion.sh` ? That makes sure that all of the various places where the version number appears stay in sync.

I'd highly recommend installing the [pre-commit](https://pre-commit.com/) hooks in your dev environment, which will do this for you :)

msp1974 replied:
Done

jeffc commented on app/app/wyoming.ts:
```diff
@@ -712,13 +721,59 @@ class WyomingServer_ {
+      const sample = dataView.getInt16(i * 2, true); // true for little-endian
+
+      // Apply gain
+      let newSample = Math.round(sample * gain * (gain * 0.6)); // roughly increasing curve from 0 to 92x
```
have you found that a linear scale makes sense in practice? or would a [reverse-?]log scale feel more natural for users?

msp1974 replied:
This is to give a more logarithimic scale.  I think some devices will benefit from a small gain, say 1-5x (setting 1 to 2) whereas some need a much bigger gain (like my Lenovo LSD) at 60-80x (8-10).

jeffc reviewed (changes requested):
Overall looks great! One small tweak requested (the `VERSION` file thing) plus a curiosity that I'll take your word for.
